Path finding optimizations. Skip (15%) closed cells early.

anvilvapre opened the pull request:
In path finding `GetConnections` considers connections to already closed cells and calculates the cost for them. The caller afterwards ignores them. These are 15% of all connections.

Add `CPos.ToOffset` to avoid `MPos` allocation in `CellLayer`.

Additional small changes with minimal performance benefit which I can imagine you won't find 'pretty'. Feel free to reject.

Use `PathGraph` directly over `IGraph` to allow non virtual index operator access. There will likely be no alternative implementations - and switching back is easy once there are.

Copies `CellInfo` to a local variable. The trade off between copying `CellInfo` versus calculating a cell index offset - and accessing an array twice.

Tested by running a replay of 20 mins.

Suggestion: remove most properties in `IGraph` and `IPathSearch` and pass them as constructor variables i.e. `PathQuery` as they are often only set once and not used outside the implementing classes.

abcdefg30 commented on OpenRA.Game/Map/CellLayer.cs:
```diff
@@ -49,12 +49,6 @@ public static CellLayer<T> CreateInstance(Func<MPos, T> initialCellValueFactory,
-		// Resolve an array index from cell coordinates
-		int Index(CPos cell)
-		{
-			return Index(cell.ToMPos(GridType));
```
Can't we do here what you currently do in `ToOffset`? I'm not sure if we want to expose that `ToOffset` logical to everything when it is only used here.

anvilvapre replied:
I first implemented it there. But then it seemed cleaner to keep all `ToMPos` logic in `CPos` - not spread duplicated code. `ToOffset` also seemed generic enough. I don't mind where.

pchote commented on OpenRA.Game/CPos.cs:
```diff
@@ -42,6 +42,7 @@ public CPos(int x, int y, byte layer)
 		}
 
 		public static readonly CPos Zero = new CPos(0, 0, 0);
+		public static readonly CPos Invalid = new CPos(-2048, -2048, 255);
```
Using magic values on CPos has caused many headaches in the past. Could we instead change `CPos IPathSearch.Expand` to `bool IPathSearch.TryExpand(out CPos)` and avoid this completely?

pchote commented:
I would support removing `IGraph` and `IPathSearch` completely. These were introduced a long time ago by someone who was trying to push for unit testing everywhere, but this did not end well and most things except for these were eventually removed.

teinarss commented on OpenRA.Mods.Common/Pathfinder/PathGraph.cs:
```diff
@@ -73,17 +73,26 @@ public GraphConnection(CPos destination, int cost)
+		public bool InReverse { get { return inReverse; } set { inReverse = value; } }
+		public Actor IgnoreActor { get { return ignoreActor; } set { ignoreActor = value; } }
+
+		// PERF: Use fields over virtual properties
+		readonly Actor actor;
+		readonly World world;
+		Func<CPos, bool> customBlock;
+		Func<CPos, int> customCost;
+		int laneBias;
+		bool inReverse;
+		Actor ignoreActor;
```
I did some testing and it looks like the jit will inline these correctly.

```cs
using System;

public interface Apa 
{
    int Prop { get; }
}

public class C : Apa {
    public int Prop { get; } = 10;
    private int field = 10;

    public int PropAccess() {
        var x = Prop + 10;
        
        return x;
    }
    
     public int FieldAccess() {
        var x = field + 10;
         
         return x;
    }
}
```

```asm
; Core CLR v5.0.421.11614 on x86

C..ctor()
    L0000: mov dword ptr [ecx+4], 0xa
    L0007: mov dword ptr [ecx+8], 0xa
    L000e: ret

C.get_Prop()
    L0000: mov eax, [ecx+4]
    L0003: ret

C.PropAccess()
    L0000: mov eax, [ecx+4]
    L0003: add eax, 0xa
    L0006: ret

C.FieldAccess()
    L0000: mov eax, [ecx+8]
    L0003: add eax, 0xa
    L0006: ret
```

anvilvapre replied:
Also in class `class D : C'? Also when there is a setter? Because with a setter i can imagine it no longer 'dares' to?
I read somewhere C# wouldn't/couldn't. Good to know.

teinarss replied:
In both those cases its able to inline it proper.

anvilvapre replied:
So shall I, some time later, remove both interfaces and change the fields back to properties? Or only the last.

teinarss replied:
Both.

teinarss commented on OpenRA.Mods.Common/Pathfinder/BasePathSearch.cs:
```diff
@@ -59,12 +59,16 @@ public interface IPathSearch : IDisposable
+
+		/// <param name="mostPromisingNode">The most promising node, if expansion was possible</param>
+		/// <returns>True if expansion was possible</returns>
+		bool Expand(out CPos mostPromisingNode);
```
Can we rename it to TryExpand?

teinarss commented on OpenRA.Mods.Common/Pathfinder/CellInfoLayerPool.cs:
```diff
@@ -15,7 +15,7 @@
 namespace OpenRA.Mods.Common.Pathfinder
 {
-	sealed class CellInfoLayerPool
+	public class CellInfoLayerPool
```
think it best to keep it sealed.

anvilvapre replied:
Not possible as Graph is public that is used by a PathSearch which is also public.

teinarss replied:
`public sealed`

teinarss commented on OpenRA.Mods.Common/Pathfinder/PathCacheStorage.cs:
```diff
@@ -9,12 +9,67 @@
+		UnitPathToRange
+	}
+
+	public struct PathCacheKey
```
this should be readonly othewise a copy will be made when using the in param, see https://docs.microsoft.com/en-us/dotnet/csharp/write-safe-efficient-code

anvilvapre commented:
Additional changes in the last push:

As requested I've refactored `BasePathSearch`, `PathSearch` from a semi builder pattern - where input and state was spread across multiple classes - to a simpler model. 

`BasePathSearch`  and `PathGraph` now take a `PathQuery` as constructor argument that defines the search to perform. Leaving only the state to be maintained in `BasePathSearch` itself. This should allow for alternative implementations more easily in future. `PathQuery` unfortunately has many fields.

While making changes it seemed to no longer make sense to keep the `PathSearch.FromPoints` and `Search` functions. `FromPoints` became `PathQueryType.Position*idirectional` and `Search` `PathQueryType.ConditionUnidirectional`.

Removed `BasePathSearch.StartsPoints` as it was not used.

`IPathSearch` and `IPathGraph` have been removed as requested. `BasePathSearch` and `PathSearch` could now also be optionally merged. Although it may make more sense to move `PathSearch.TryExpand` to `PathGraph` rather than `BasePathSearch`.

Pulled some of the calculation in some `CustomCost` functions for path searching outside the functions.

Note that by not adding `self.Location` twice as start point for path searches results in rare conditions in a slightly different path. Which impacts Sync. This is correct.

No functional changes (intended). Apart from the impact on sync mentioned above.

Tested by running a 7 minute replay that was recorded on bleed (with the one change that impacts sync temporary disabled).

teinarss commented on OpenRA.Mods.Common/Activities/FindAndDeliverResources.cs:
```diff
@@ -162,12 +162,14 @@ public override bool Tick(Actor self)
 			{
-				if (harv.CanHarvestCell(self, self.Location) && claimLayer.CanClaimCell(self, self.Location))
+				if (harv.CanHarvestCell(self, self.Location)
+						&& claimLayer.CanClaimCell(self, self.Location))
```
Only formatting changes?

teinarss commented on OpenRA.Mods.Common/Activities/FindAndDeliverResources.cs:
```diff
@@ -162,12 +162,14 @@ public override bool Tick(Actor self)
 			{
-				if (harv.CanHarvestCell(self, orderLocation.Value) && claimLayer.CanClaimCell(self, orderLocation.Value))
+				if (harv.CanHarvestCell(self, orderLocation.Value)
+						&& claimLayer.CanClaimCell(self, orderLocation.Value))
```
Same here

teinarss commented on OpenRA.Mods.Common/Traits/Harvester.cs:
```diff
@@ -178,38 +178,58 @@ bool IsAcceptableProcType(Actor proc)
+			var refineries = world.ActorsWithTrait<IAcceptResources>()
+				.Where(r => r.Actor != ignore
+					&& r.Actor.Owner == self.Owner
+					&& IsAcceptableProcType(r.Actor))
```
Please keep this in one line.

reaperrr commented on .github/workflows/ci.yaml:
```diff
@@ -64,7 +64,7 @@ jobs:
         run: |
           # Work around runtime failures on the GH Actions runner
-          dotnet nuget locals all --clear
+          dotnet nuget add source https://api.nuget.org/v3/index.json -n nuget.org
```
I don't think this belongs in this PR?

anvilvapre replied:
Taken from #19325 to be able to let the Windows build succeed after recent changes on the github side.

pchote suggested to wait for that pull request to be merged.

teinarss commented on OpenRA.Mods.Common/Traits/World/ActorMap.cs:
```diff
@@ -261,7 +261,8 @@ public IEnumerable<Actor> GetActorsAt(CPos a)
 				return Enumerable.Empty<Actor>();
 
-			var layer = a.Layer == 0 ? influence : customInfluence[a.Layer];
+			var layerIndex = a.Layer;
+			var layer = layerIndex == 0 ? influence : customInfluence[layerIndex];
```
Fun fact: It did look like the jit compiler was able to do this optimization by itself.

pchote replied:
Should we drop these then?

pchote commented on OpenRA.Game/CPos.cs:
```diff
@@ -85,6 +85,11 @@ public MPos ToMPos(MapGridType gridType)
 			return new MPos(u, v);
 		}
 
+		public CPos AtLayer(byte layer)
```
Can we please call this `ToLayer`?

anvilvapre replied:
Yes. But it will look similar to `ToMPos`.

pchote replied:
Ah, thats a good point. How about `WithLayer` then.

`AtLayer` sounds like it should be returning true/false if the given layer matches...

pchote commented on OpenRA.Game/Primitives/PriorityQueue.cs:
```diff
@@ -41,10 +43,11 @@ public void Add(T item)
+			T above;
 
-			while (addLevel >= 1 && comparer.Compare(Above(addLevel, addIndex), item) > 0)
+			while (addLevel >= 1 && comparer.Compare((above = Above(addLevel, addIndex)), item) > 0)
```
We generally try to avoid using inline assignments (not sure the correct term) like this. Can we restructure the loop so that above is assigned on its own line?

pchote commented on OpenRA.Game/Primitives/PriorityQueue.cs:
```diff
@@ -91,6 +94,33 @@ public T Pop()
+		{
+			if (level == 0)
+			{
+				item = default(T);
```
This can now be just `item = default;`, likewise in `TryPop`.

pchote commented on OpenRA.Game/Primitives/PriorityQueue.cs:
```diff
@@ -91,6 +94,33 @@ public T Pop()
+
+		public bool TryPop(out T item)
+		{
+			if (level == 0)
```
`Pop` also checks `index`. Is it safe to ignore that here?

anvilvapre replied:
`Empty` only uses `level`. So I assumed it would be safe.

pchote reviewed (commented):
A couple of small comments i noticed while skimming the commit below.

My overall feedback is that this is trying to do way too many different things within a single commit. This makes it very difficult to understand how the different changes relate to eachother and overall difficult to review. Can you please split this up into individual commits (they can all still be in this PR) for each logicial change?

e.g. introducing `PathQuery` should have its own commit that switches the code to the new class without changing any behaviour, the harvester pathfinding improvements should have another, the different micro-optimizations each get their own, and so on.

anvilvapre committed:
Fix github Windows build issue.

anvilvapre committed:
CPos.ToLayer.

anvilvapre committed:
CellLayer inline CPos.ToMPos.

anvilvapre committed:
PathCache use PathCacheKey over string as key.

anvilvapre committed:
PriorityQueue, TryPeek, TryPop.

anvilvapre committed:
Do no lookup cell layer twice.

anvilvapre committed:
Avoid unnecessary cast.

pchote commented:
Can you please try to add the changes that use `ToLayer`, `TryPeek`, `TryPop` into the same commits that add them? The value in splitting things into different commits is that we can see each set of changes in isolation and the evolution of the code, rather than have a diff that simply swaps one implementation for another.

anvilvapre commented:
> Can you please try to add the changes that use `ToLayer`, `TryPeek`, `TryPop` into the same commits that add them? The value in splitting things into different commits is that we can see each set of changes in isolation and the evolution of the code, rather than have a diff that simply swaps one implementation for another.

I basically can't. I rebased everything quite often because I thought you wanted it. I made changes before and after committing PathQuery and they spread multiple files. After i renamed PathQuery where it first had a different name. Evolution took side paths ; )

I agree PathQuery could have better been in a separate pull request. Should have suggested this.

Let me one day find the patience to reapply all changes manually across files and turn them into multiple pull requests as well.

anvilvapre commented:
Current split up seems the most fine grained I can make it without having to rewrite code and unravel multiple changes per file even per line. Started with a mergetool but it really means rewriting all changes manually.

Note that `CPos.WithLayer` as a change is not related to `PriorityQueue.TryPop`. They are to optimize `PathGraph` and `PathSearch` independantly.

pchote commented:
> Let me one day find the patience to reapply all changes manually across files and turn them into multiple pull requests as well.

This shouldn't be necessary!

Here is the workflow I usually use in this situation, which is faster in practice than my list might imply:

1. Check out a duplicate copy of the branch (e.g. `pathfinder-temp`
2. Decide on what subset of changes I want to go in the first commit
3. Reset all changes that aren't related to those changes (using a graphical tool like gitg makes it a *lot* easier to unstage individual lines)
4. Fix any compile errors and test that the subset of features works as intended
5. Create the first commit
6. Switch back to the first branch and rebase on the temp branch. Resolve any rebase issues and if needed test anything that you suspect might have regressed.
7. Repeat steps 1-5 as many times as needed.

A lot of the changes end up being obviously distinct, which means that you can make several commits in a single pass. Its only the core parts where different features overlap where the full loop is needed.

anvilvapre committed:
Split PathSearch in PathQuery and PathSearch.

anvilvapre committed:
PriorityQueue, minor optimizatoin.

anvilvapre commented:
> A lot of the changes end up being obviously distinct, which means that you can make several commits in a single pass. Its only the core parts where different features overlap where the full loop is needed.

It was the process that i started.  I've extracted the distinct. Because I changed a lot of class field members I can no longer easily pick distinct changes without breaking the code. 

It means that the last commit now contains three changes: `PathQuery`, a one liner for skipping 15% of closed cells early and some optimizations for custom goals functions that were passed to path searches.

AspectInteractive commented:
> `IPathSearch` and `IPathGraph` have been removed as requested. 

This seems reasonable since BasePathSearch is the only class that uses these interfaces.

> `BasePathSearch` and `PathSearch` could now also be optionally merged. 

I am quite new and not that familiar with the code base, but would this not prevent alternative PathSearch implementations from being used? I can understand removing the interface, as it has many implementation details that are unlikely to be used by a second class, however BasePathSearch appears general enough - or at least should be - to take several implementations. PathSearch directly implements the A* Algorithm, meaning all pathfinding would need to use this A* algorithm from this point onwards. Given that pathfinding is still not an entirely solved problem, I don't think hardcoding the A* algorithm here (or any specific pathfinding algorithm) is a wise choice.

pchote commented:
Activities are currently hardcoding `PathSearch` directly, so even with all the interfaces it wasn't possible to switch out any implementations. *If we want to fix this*, this can be done by adding path search wrappers to `IPathfinder` and changing activities to call these instead.

I'm not convinced we win much by fixing this, however, since these activities all make strong assumptions about the cell and pathfinding system and likely need to be completely replaced by any mod that wants to change the way the grid movement system works.

AspectInteractive commented:
> I'm not convinced we win much by fixing this, however, since these activities all make strong assumptions about the cell and pathfinding system and likely need to be completely replaced by any mod that wants to change the way the grid movement system works.

There are still other grid-based path finders you may wish to use in the future. A basic implementation of the flow field algorithm for instance is compatible with grid-based movement. For this reason I agree with your initial point of fixing the iPathFinder interface (and eventually decoupling Activities) rather than removing it entirely.

penev92 commented:
I believe this is the original/base PR that is getting split over multiple others with the intent of getting closed in the end so to mitigate (my own) confusion I'll convert this into a Draft PR for the time being.

anvilvapre commented:
Closing to reduce number of open PRs. This pull request was split up in many smaller ones. Some merged, some waiting for approval.
